pkg/tfbridge/token.go

```diff
@@ -189,3 +191,169 @@ func (err UnmappableError) Error() string {
 func (err UnmappableError) Unwrap() error {
 	return err.Reason
 }
+
+type tokenHistory[T ~string] struct {
+	Current T          `json:"current"`        // the current Pulumi token for the resource
+	Past    []alias[T] `json:"past,omitempty"` // Previous tokens
+}
+
+type alias[T ~string] struct {
+	Name      T    `json:"name"`      // The previous token.
+	InCodegen bool `json:"inCodegen"` // If the alias is a fully generated resource, or just a schema alias.
+}
+
+type aliasHistory struct {
+	Resources   map[string]*tokenHistory[tokens.Type]         `json:"resources"`
+	DataSources map[string]*tokenHistory[tokens.ModuleMember] `json:"datasources"`
```

Comment:
Interesting, so this data is not stored in schema? Is it stored in MarshallableProviderInfo? Perhaps a dedicated struct is better anyway.

Reply:
The alias history isn't really part of the ProviderInfo or the schema. It is a historical artifact of the mapping between the two.

Reply:
Got it, the coin dropped. Yes it needs an accumulating history.